File: internal/model/incident.go
// Package model defines the core data structures shared across Nexus.
//
// I keep these free of behaviour beyond simple helpers: probes,
// analyzers, remediators, and stores all communicate through these
// types, so they must stay dependency-free.
package model

import (
	"encoding/json"
	"time"
)

// Incident is a detected problem moving through the closed loop.
type Incident struct {
	ID       string         `json:"id" db:"id"`
	Type     IncidentType   `json:"type" db:"type"`
	Severity Severity       `json:"severity" db:"severity"`
	Status   IncidentStatus `json:"status" db:"status"`

	// Source
	ProbeID      string `json:"probe_id" db:"probe_id"`
	TargetID     string `json:"target_id" db:"target_id"`
	SourceSignal string `json:"source_signal" db:"source_signal"`

	// Timeline
	DetectedAt  time.Time  `json:"detected_at" db:"detected_at"`
	DiagnosedAt *time.Time `json:"diagnosed_at,omitempty" db:"diagnosed_at"`
	FixedAt     *time.Time `json:"fixed_at,omitempty" db:"fixed_at"`
	VerifiedAt  *time.Time `json:"verified_at,omitempty" db:"verified_at"`
	ResolvedAt  *time.Time `json:"resolved_at,omitempty" db:"resolved_at"`

	// Diagnosis
	RootCause  string  `json:"root_cause" db:"root_cause"`
	Confidence float64 `json:"confidence" db:"confidence"` // 0.0 to 1.0

	// Fix
	FixGenerated bool   `json:"fix_generated" db:"fix_generated"`
	FixPR        string `json:"fix_pr_url" db:"fix_pr_url"`
	FixBranch    string `json:"fix_branch" db:"fix_branch"`
	FixSummary   string `json:"fix_summary" db:"fix_summary"`

	// Verification
	Verified    *bool `json:"verified,omitempty" db:"verified"`
	MTTRSeconds int64 `json:"mttr_seconds" db:"mttr_seconds"`

	// Audit
	CycleID  string          `json:"cycle_id" db:"cycle_id"`
	Log      json.RawMessage `json:"log,omitempty" db:"log"`
	Metadata map[string]any  `json:"metadata,omitempty" db:"metadata"`
}

// IncidentType classifies what kind of problem was detected.
type IncidentType string

// Incident type values recognised by the analyzers.
const (
	IncidentCostSpike       IncidentType = "cost_spike"
	IncidentPerformance     IncidentType = "performance_degradation"
	IncidentSecurityDrift   IncidentType = "security_drift"
	IncidentComplianceDrift IncidentType = "compliance_drift"
	IncidentReliability     IncidentType = "reliability_degradation"
	IncidentScale           IncidentType = "scaling_bottleneck"
	IncidentConfigDrift     IncidentType = "configuration_drift"
	IncidentResourceExhaust IncidentType = "resource_exhaustion"
	IncidentErrorBurst      IncidentType = "error_burst"
	IncidentCustom          IncidentType = "custom"
)

// Severity ranks how urgent an incident is.
type Severity string

// Severity values ordered from most to least urgent.
const (
	SeverityCritical Severity = "critical"
	SeverityHigh     Severity = "high"
	SeverityMedium   Severity = "medium"
	SeverityLow      Severity = "low"
	SeverityInfo     Severity = "info"
)

// IncidentStatus tracks progress through the closed loop.
type IncidentStatus string

// Incident status values covering the closed-loop lifecycle.
const (
	StatusDetected   IncidentStatus = "detected"
	StatusDiagnosing IncidentStatus = "diagnosing"
	StatusDiagnosed  IncidentStatus = "diagnosed"
	StatusFixing     IncidentStatus = "fixing"
	StatusFixReady   IncidentStatus = "fix_ready" // PR open, awaiting approval
	StatusApplying   IncidentStatus = "applying"
	StatusVerifying  IncidentStatus = "verifying"
	StatusResolved   IncidentStatus = "resolved"
	StatusFailed     IncidentStatus = "failed"
	StatusEscalated  IncidentStatus = "escalated" // blocked by a policy gate
)

// validTransitions encodes the incident state machine. I enforce this
// centrally so no component can move an incident into an illegal state.
var validTransitions = map[IncidentStatus][]IncidentStatus{
	StatusDetected:   {StatusDiagnosing, StatusFailed, StatusEscalated},
	StatusDiagnosing: {StatusDiagnosed, StatusFailed, StatusEscalated},
	StatusDiagnosed:  {StatusFixing, StatusFailed, StatusEscalated},
	StatusFixing:     {StatusFixReady, StatusFailed, StatusEscalated},
	StatusFixReady:   {StatusApplying, StatusFailed, StatusEscalated},
	StatusApplying:   {StatusVerifying, StatusFailed, StatusEscalated},
	StatusVerifying:  {StatusResolved, StatusFailed, StatusEscalated},
	StatusResolved:   {},
	StatusFailed:     {StatusDiagnosing}, // allow a retry from diagnosis
	StatusEscalated:  {StatusFixReady, StatusFailed},
}

// CanTransition reports whether an incident may move from its current
// status to the requested one.
func (i *Incident) CanTransition(to IncidentStatus) bool {
	for _, allowed := range validTransitions[i.Status] {
		if allowed == to {
			return true
		}
	}
	return false
}

File: nexus/models/incident.py
from __future__ import annotations
from datetime import datetime
from enum import Enum
from typing import Any, Dict, List, Optional
from uuid import uuid4
from pydantic import BaseModel, Field
class IncidentType(str, Enum):
    cost_spike = "cost_spike"
    performance_degradation = "performance_degradation"
    security_drift = "security_drift"
    compliance_drift = "compliance_drift"
    reliability_degradation = "reliability_degradation"
    scaling_bottleneck = "scaling_bottleneck"
    configuration_drift = "configuration_drift"
    resource_exhaustion = "resource_exhaustion"
    error_burst = "error_burst"
    custom = "custom"
class Severity(str, Enum):
    critical = "critical"; high = "high"; medium = "medium"; low = "low"; info = "info"
class IncidentStatus(str, Enum):
    detected = "detected"; diagnosing = "diagnosing"; diagnosed = "diagnosed"; fixing = "fixing"; fix_ready = "fix_ready"; applying = "applying"; verifying = "verifying"; resolved = "resolved"; failed = "failed"; escalated = "escalated"
VALID_TRANSITIONS: Dict[IncidentStatus, List[IncidentStatus]] = {
    IncidentStatus.detected: [IncidentStatus.diagnosing, IncidentStatus.failed, IncidentStatus.escalated],
    IncidentStatus.diagnosing: [IncidentStatus.diagnosed, IncidentStatus.failed, IncidentStatus.escalated],
    IncidentStatus.diagnosed: [IncidentStatus.fixing, IncidentStatus.failed, IncidentStatus.escalated],
    IncidentStatus.fixing: [IncidentStatus.fix_ready, IncidentStatus.failed, IncidentStatus.escalated],
    IncidentStatus.fix_ready: [IncidentStatus.applying, IncidentStatus.failed, IncidentStatus.escalated],
    IncidentStatus.applying: [IncidentStatus.verifying, IncidentStatus.failed, IncidentStatus.escalated],
    IncidentStatus.verifying: [IncidentStatus.resolved, IncidentStatus.failed, IncidentStatus.escalated],
    IncidentStatus.resolved: [],
    IncidentStatus.failed: [IncidentStatus.diagnosing],
    IncidentStatus.escalated: [IncidentStatus.fix_ready, IncidentStatus.failed],
}
class Incident(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid4()))
    type: IncidentType
    severity: Severity
    status: IncidentStatus = IncidentStatus.detected
    probe_id: str = ""
    target_id: str = ""
    source_signal: str = ""
    detected_at: datetime = Field(default_factory=datetime.utcnow)
    diagnosed_at: Optional[datetime] = None
    fixed_at: Optional[datetime] = None
    verified_at: Optional[datetime] = None
    resolved_at: Optional[datetime] = None
    root_cause: str = ""
    confidence: float = Field(default=0.0, ge=0.0, le=1.0)
    fix_generated: bool = False
    fix_pr_url: str = ""
    fix_branch: str = ""
    fix_summary: str = ""
    verified: Optional[bool] = None
    mttr_seconds: int = 0
    cycle_id: str = ""
    log: List[Dict[str, Any]] = Field(default_factory=list)
    metadata: Dict[str, Any] = Field(default_factory=dict)
    def can_transition(self, to: IncidentStatus) -> bool:
        return to in VALID_TRANSITIONS.get(self.status, [])
    def transition(self, to: IncidentStatus) -> None:
        if not self.can_transition(to):
            raise ValueError(f"illegal status transition {self.status} -> {to} for incident {self.id}")
        self.status = to
        now = datetime.utcnow()
        if to == IncidentStatus.diagnosed: self.diagnosed_at = now
        elif to == IncidentStatus.fix_ready: self.fixed_at = now
        elif to == IncidentStatus.verifying: self.verified_at = now
        elif to == IncidentStatus.resolved:
            self.resolved_at = now
            if self.detected_at: self.mttr_seconds = int((now - self.detected_at).total_seconds())
